angular/src/lib/select/select.component.ts

```diff
@@ -115,306 +226,474 @@ export class SelectComponent implements OnInit, AfterContentChecked, AfterConten
-        this._keyManager.setActiveItem(matchedOption);
+      if (rowDataKeyValue) {
+        this.selectionKeys[rowDataKeyValue] = 1;
       }
     }
+    this.selectService.onSelectionChange();
   }
 
-  private _selectValue = (value: any): ListItemComponent | undefined => {
-    const matchedOption = this.selectOptions.find((option: ListItemComponent) => {
-      try {
-        return option.label != null && this._compareWith(option.label, value);
-      } catch (error) {
-        console.warn(error);
+  updateSelectionKeys() {
+    if (this.dataKey && this._selection) {
+
+      this.selectionKeys = {};
+      if (Array.isArray(this._selection)) {
+        for (const data of this._selection) {
+          this.selectionKeys[String(data[this.dataKey])] = 1;
+        }
+      } else {
+        this.selectionKeys[String(this._selection[this.dataKey])] = 1;
       }
-      return false;
-    });
+    }
+  }
 
-    if (matchedOption) {
-      this._selectionModel.select(matchedOption);
+  constructor(
+    public selectService: SelectService,
+    public el: ElementRef,
+    private cd: ChangeDetectorRef
+  ) { }
+
+  ngAfterContentChecked () {
+    this._setAnchorWidth(this.originButton.el.nativeElement);
+  }
+
+  toggle = (): void => {
+    this.overlayOpen ? this.close() : this.open();
+  }
+
+  open = (): void => {
+    if (!this.options) {
+      return;
+    }
+    this.handleShow.emit();
+    this.overlayOpen = true;
+    this.cd.detectChanges();
+
+    if (this.filterViewChild && this.filterViewChild.nativeElement) {
+      this.filterViewChild.nativeElement.focus();
     }
+  }
 
-    return matchedOption;
+  close = (): void => {
+    if (this.overlayOpen) {
+      this.overlayOpen = false;
+    }
   }
 
-  private _propagateChanges = (fallbackValue?: any): void => {
-    let valueToEmit: any = null;
+  onModelChange: Function = () => {};
+  onModelTouched: Function = () => {};
 
-    if (this.isMulti) {
-      valueToEmit = (this.selected as ListItemComponent[]).map(option => option.label);
-    } else {
-      valueToEmit = this.selected ? (this.selected as ListItemComponent).label : fallbackValue;
+  writeValue(value: any): void {
+    if (this.filter) {
+      this.resetFilter();
     }
+    this.value = value;
+    this.updateSelectedItem(value);
+    this.cd.markForCheck();
+  }
+  registerOnChange(fn: Function): void {
+    this.onModelChange = fn;
+  }
+  registerOnTouched(fn: Function): void {
+    this.onModelTouched = fn;
+  }
+  setDisabledState(isDisabled: boolean): void {
+    this.disabled = isDisabled;
+  }
 
-    this._ngModelValue = valueToEmit;
-    this.ngModelChange.emit(valueToEmit);
-    this._onChange(valueToEmit);
-    this.selectionChange.emit(new SelectChange(this, valueToEmit));
+  private _setAnchorWidth = (elementAnchor) => {
+    const anchor = elementAnchor && elementAnchor.getBoundingClientRect();
+    this.anchorWidth = anchor.width;
   }
+}
+
+@Component({
+  selector: 'md-selectItem',
```

Comment:
Not sure I agree with this naming for a component.  Is there a reason you went camelcase for a component?  It also makes the scss unconventional.

Reply:
good catch, switching to md-select-item